### DISCUSSION TITLE
Alpine not removing classes on transition end.

### DISCUSSION THREAD

**Discussion by iivvaannxx** [alpinejs/alpine#3310] [7. Bugs (archived)] [answered]
Okay, this is quite a weird bug, I think.

## The bug

I have a component (defined with Alpine.data) CrossfadeCarousel which, as its name says, displays a slideshow of given images using a crossfade transition. It has the following properties:

```typescript
export default {

    /** @brief The duration of the change animation. */
    'duration': 500,

    /** @brief The interval at which the image of the carousel gets updated. */
    'interval': 5000,


    /** @brief The index of the currently displayed image. */
    'current': 1,
}
```

Basically, when the component initializes (through the 'init' callback) I set an interval which updates the 'current' property. Then the childs of the 'carousel' component are another kind of Alpine components which I defined as 'Carousel Targets' and they have the following configuration.

```typescript
export default function (this: AlpineDataContext <any>) {

    function initialize (this: AlpineDataContext <any>) {

        this.key = this.$id("slide");
    }

    function isVisible (this: AlpineDataContext <any>) {

        return `slide-${this.current}` === this.key;
    }

    return {

        init: initialize,
        visible: isVisible,

        key: 'x',
        fade: animatedShowIf('visible', {

            'enter': `transition-opacity ease-out duration-2000 !absolute left-0 top-0`,
            'enter-start': 'opacity-0',
            'enter-end': 'opacity-100',

            'leave': `transition-opacity ease-in duration-2000 !relative`,
            'leave-start': 'opacity-100',
            'leave-end': 'opacity-0',
        }),
    }
}
```

Forget about the parameters of the functions, they are only used to make Typescript recognize the `this` context when inside those functions. Basically, each target gets an ID automatically assigned, and that is used to compare with the `current` property, and if they match, the `current` target is shown. This is made by the function `animatedShowIf` which binds the `x-show` directive to the given conditional element (in this case the `visible` property), and also defines which transitions use.

This works like a charm, and the transition can be completely regulated by parameters of the component (not with this code, as it's the configuration I am using to debug its behaviour). The thing comes when I switch the active browser tab. I leave the website for a little while and the interval keeps running and so, the images keep changing. After a little time, when I come back, all the images are stacked up one over the other and all the `picture` elements which 'transitioned' while I was on another tab have the 'leave' classes still applied. 

## Current behaviour
See the image below, the classes are applied to 4 different elements, while theoretically (and it's how it works when I am focusing the tab) only two elements must be transitioning at any given moment (one with the !absolute position, and another with the !relative position, in case you want to distinguish them). In this image, the element with `key = 3` is transitioning over the element with `key = 2` and all the others should not be visible and have their respective `display: none` attached by the `x-show` directive.

![image](https://user-images.githubusercontent.com/47715589/205183859-65ae8250-bf5c-44b1-a00d-071597f26ea7.png)

After I return to the page, they almost disappear instantly (while actually doing the transition) but I don't think that would be the expected behaviour. Here I am attaching a video to show the problem.

https://user-images.githubusercontent.com/47715589/205183794-20a10577-1bfb-4045-a693-8e5c21c25e98.mp4

I suppose that one solution could be to stop the interval if I leave the tab (I am not sure if that can be done but I am guessing that there sould be some way). But I am reporting the issue in case it could lead to some bugfix (and hoping it's a bug of the framework and not a problem of mine).

## Expected behaviour
The classes get removed and the elements are set to `display: none` even if I am not focusing the tab.

## Additional Information
 - Alpine Version: v3.10.5
 - Browser: Microsoft Edge

 - This project it's being made with [Astro](https://astro.build) and [Tailwind CSS](https://tailwindcss.com)

**Comment by ekwoka** [comment, marked as answer]
I'm not sure if this is in the scope of Alpine.

When a tab with timers is unfocussed the browser often fudges the timers quite heavily and will execute a bunch of them right on return to focus, and Alpine is just handling those as they happen.

Similarly do css animations and transitions run when noone is around to see them?

I guess Alpine could recognize the tab is hidden and just skip over transitions entirely.

**Reply by iivvaannxx** [reply]
I am actually a little ashamed I didn't think of something as trivial as that. Certainly, it would be great if Alpine could do that, but I think that it would be more of an edge case than an actual feature. Thanks! I will fix it by simply pausing the transitions using the blur event or something like that.
